src/nginx/t/metadata_fecth_fail.t

```diff
@@ -0,0 +1,219 @@
+
+use src::nginx::t::ApiManager; # Must be first (sets up import path to the Nginx test module)
+use src::nginx::t::HttpServer;
+use src::nginx::t::ServiceControl;
+use Test::Nginx; # Imports Nginx's test module
+use Test::More;  # And the test framework
+
+################################################################################
+
+# Port assignments
+my $NginxPort = ApiManager::pick_port();
+my $BackendPort = ApiManager::pick_port();
+my $ServiceControlPort = ApiManager::pick_port();
+my $MetadataPort = ApiManager::pick_port();
+
+my $t = Test::Nginx->new()->has(qw/http proxy/)->plan(16);
+
+my $config = ApiManager::get_bookstore_service_config . <<"EOF";
+control {
+  environment: "http://127.0.0.1:${ServiceControlPort}"
+}
+EOF
+
+$t->write_file('service.pb.txt', $config);
+ApiManager::write_file_expand($t, 'nginx.conf', <<"EOF");
+%%TEST_GLOBALS%%
+daemon off;
+events { worker_connections 32; }
+http {
+  %%TEST_GLOBALS_HTTP%%
+  server_tokens off;
+  endpoints {
+    metadata_server http://127.0.0.1:${MetadataPort};
+  }
+  server {
+    listen 127.0.0.1:${NginxPort};
+    server_name localhost;
+    location / {
+      endpoints {
+        api service.pb.txt;
+        %%TEST_CONFIG%%
+        on;
+      }
+      proxy_pass http://127.0.0.1:${BackendPort};
+    }
+  }
+}
+EOF
+
+sub no_check_call {
+    my (@requests) = @_;
+    foreach my $r (@requests) {
+        if ($r->{path} =~ qr/:check$/) {
+            return 0;
+        }
+    }
+    return 1;
+}
+
+sub test_metadata {
+    my ($sleep, $wantReqHeader, $wantReqBody) = @_;
+    my $report_done = 'report_done';
+    my $backend_log = 'backend.log';
+    my $servicecontrol_log = 'servicecontrol.log';
+
+    $t->run_daemon(\&backends, $t, $BackendPort, $backend_log);
+    $t->run_daemon(\&servicecontrol, $t, $ServiceControlPort, $servicecontrol_log, $report_done);
+    $t->run_daemon(\&metadata, $t, $MetadataPort, 'metadata.log');
+
+    is($t->waitforsocket("127.0.0.1:${BackendPort}"), 1, 'bookstore socket ready.');
+    is($t->waitforsocket("127.0.0.1:${ServiceControlPort}"), 1, 'Service control socket ready.');
+    is($t->waitforsocket("127.0.0.1:${MetadataPort}"), 1, 'Metadata socket ready.');
+
+    $t->run();
+
+    ################################################################################
+
+    my $shelves1 = ApiManager::http_get($NginxPort, '/shelves');
+    is($t->waitforfile("$t->{_testdir}/${report_done}"), 1, 'Report succeeded');
+
+    my ($shelves_headers1, $shelves_body1) = split /\r\n\r\n/, $shelves1, 2;
+    like($shelves_headers1, qr/HTTP\/1\.1 500 Internal Server Error/, '/shelves returned HTTP 500.');
+    like($shelves_body1, qr/Failed to fetch service account token/, 'Proxy failed in fetch service account token');
+
+    ################################################################################
+    # if no sleep, the service account token still doesn't get out of last failed
+    # fetch so it will fail.
+    if ($sleep) {
```

Comment:
how about instead of boolean, passing in sleep time in second.   first test is 2s, which will fail,  second test is 6s, the secnod call should work